source/dub/packagemanager.d

```diff
@@ -445,138 +494,126 @@ class PackageManager {
-		foreach (p; *packs) {
-			if (p.path == path) {
-				enforce(p.version_ == pack.version_, "Adding the same local package twice with differing versions is not allowed.");
-				logInfo("Package is already registered: %s (version: %s)", p.name, p.version_);
-				return p;
-			}
-		}
-
-		addPackages(*packs, pack);
-
-		writeLocalPackageList(type);
-
-		logInfo("Registered package: %s (version: %s)", pack.name, pack.version_);
-		return pack;
+		return addLocalPackage(path, verName, m_defaultPaths[type]);
+	}
+	Package addLocalPackage(Path path, string verName, Path repoPath)
+	in { assert(repoPath.absolute); }
+	body
+{
+		repoPath.normalize();
+		return m_repos[repoPath].addLocalPackage(path, verName);
 	}
 
-	void removeLocalPackage(Path path, LocalPackageType type)
+	deprecated void removeLocalPackage(Path path, LocalPackageType type)
 	{
-		path.endsWithSlash = true;
-
-		Package[]* packs = &m_repositories[type].localPackages;
-		size_t[] to_remove;
-		foreach( i, entry; *packs )
-			if( entry.path == path )
-				to_remove ~= i;
-		enforce(to_remove.length > 0, "No "~type.to!string()~" package found at "~path.toNativeString());
-
-		string[Version] removed;
-		foreach_reverse( i; to_remove ) {
-			removed[(*packs)[i].version_] = (*packs)[i].name;
-			*packs = (*packs)[0 .. i] ~ (*packs)[i+1 .. $];
-		}
-
-		writeLocalPackageList(type);
-
-		foreach(ver, name; removed)
-			logInfo("Deregistered package: %s (version: %s)", name, ver);
+		removeLocalPackage(path, m_defaultPaths[type]);
+	}
+	void removeLocalPackage(Path path, Path repoPath)
+	in { assert(repoPath.absolute); }
+	body
+	{
+		repoPath.normalize();
+		m_repos[repoPath].removeLocalPackage(path);
 	}
 
 	/// For the given type add another path where packages will be looked up.
-	void addSearchPath(Path path, LocalPackageType type)
+	deprecated void addSearchPath(Path path, LocalPackageType type)
 	{
-		m_repositories[type].searchPath ~= path;
-		writeLocalPackageList(type);
+		addSearchPath(path, m_defaultPaths[type]);
+	}
+	/// ditto
+	void addSearchPath(Path path, Path repoPath)
+	in { assert(repoPath.absolute); }
+	body
+	{
+		repoPath.normalize();
+		m_repos[repoPath].addSearchPath(path);
 	}
 
 	/// Removes a search path from the given type.
-	void removeSearchPath(Path path, LocalPackageType type)
+	deprecated void removeSearchPath(Path path, LocalPackageType type)
+	{
+		removeSearchPath(path, m_defaultPaths[type]);
+	}
+	/// ditto
+	void removeSearchPath(Path path, Path repoPath)
+	in { assert(repoPath.absolute); }
+	body
 	{
-		m_repositories[type].searchPath = m_repositories[type].searchPath.filter!(p => p != path)().array();
-		writeLocalPackageList(type);
+		repoPath.normalize();
+		m_repos[repoPath].removeSearchPath(path);
 	}
 
 	void refresh(bool refresh_existing_packages)
 	{
 		logDiagnostic("Refreshing local packages (refresh existing: %s)...", refresh_existing_packages);
 
 		// load locally defined packages
-		void scanLocalPackages(LocalPackageType type)
```

Comment:
There's a big diff here because of adding a level of indent. See individual comments for actual changes.